core/src/main/kotlin/org/evomaster/core/problem/api/service/ApiWsStructureMutator.kt

```diff
@@ -25,56 +26,134 @@ import kotlin.math.min
 
         if (config.externalServiceIPSelectionStrategy == EMConfig.ExternalServiceIPSelectionStrategy.NONE) {
             return
         }
 
         if (sampler is ResourceSampler) {
-            val ind = individual.individual as? T
+            val ind = individual.individual as? ApiWsIndividual
                 ?: throw IllegalArgumentException("Invalid individual type")
 
-            // TODO: Under review
-            val fw = individual.fitness.getAccessedExternalServiceRequests()
+            val esr = individual.fitness.getViewAccessedExternalServiceRequests()
+            if (esr.isEmpty()) {
+                //nothing to do
+                return
+            }
 
-            val actions = mutableListOf<ExternalServiceAction>().plus(
-                sampler.getExternalService().getExternalServiceActions()
-            )
+            ind.seeMainExecutableActions().forEachIndexed { index, action ->
+
+                val parent = action.parent
+                if (parent !is EnterpriseActionGroup) {
+                    //TODO this should not really happen
+                    val msg = "Action is not inside an EnterpriseActionGroup"
+                    log.error(msg)
+                    throw RuntimeException(msg)
+                }
+
+                // If the action already exists mark it as used
+                esr.flatMap { it.value }.forEach { url ->
+                    val existingActions = parent.getViewOfChildren()
+                        .filterIsInstance<HttpExternalServiceAction>()
+                        .filter { it.request.absoluteURL == url }
+
+                    val actions = sampler.getExternalService()
+                        .getExternalServiceActions()
+                        .filter { it.request.absoluteURL == url }
+
+                    if (existingActions.isEmpty()) {
+                        if (actions.isNotEmpty()) {
+                            parent.addChildrenToGroup(
+                                actions,
+                                GroupsOfChildren.EXTERNAL_SERVICES
+                            )
+
+                            // update impact based on added genes
+                            // TODO: Man to review this..
+                            if (mutatedGenes != null && actions.isNotEmpty() && config.isEnabledArchiveGeneSelection()) {
+                                individual.updateImpactGeneDueToAddedExternalService(mutatedGenes, actions)
+                            }
+                        }
+                    } else {
+                        /*
+                            nothing was called so anything there from previous setups can be safely removed.
+                            however, "removing" could have some side-effects (TODO explains), so we just mark them
+                            as disabled
+
+                            Code never reaches this part
+                        */
+                        parent.getViewOfChildren()
```

Comment:
As we discussed, we do not `confirmNotUsed` here, we did it after fitness evaluation.